docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+be waiting for all dependencies to update synchronously. For instance, we could
+have 3 versions of the core SDK compatible with the latest 2 releases of
+CosmWasm as well as 4 different versions of staking . This sort of setup would
+allow early adopters to aggressively integrate new versions, while allowing 
+more conservative users to be selective about which versions they're ready for.
+
+In order to achieve this, we need to solve the following problems:
+1. because of the way [go semantic import versioning](https://research.swtch.com/vgo-import) (SIV)
+works, moving to SIV naively will actually make it harder to achieve these goals
+2. circular dependencies between modules need to be broken to actually release
+many modules in the SDK independently
+3. pernicious minor version incompatibilities introduced through correctly
+[evolving protobuf schemas](https://developers.google.com/protocol-buffers/docs/proto3#updating)
+without correct [unknown field filtering](./adr-020-protobuf-transaction-encoding.md#unknown-field-filtering)
+
+Note that all the following discussion assumes that the proto file versioning and state machine versioning of a module
+are distinct in that:
+* proto files are maintained in a non-breaking way (using something like [buf breaking](https://docs.buf.build/breaking/overview)
+  to ensure all changes are backwards compatible)
+* proto file versions get bumped much less frequently, i.e. we might maintain `cosmos.bank.v1` through many versions
+  of the bank module state machine
+* state machine breaking changes are more common and ideally this is what we'd want to semantically version with
+go modules, ex. `x/bank/v2`, `x/bank/v3`, etc.
+
+### Problem 1: Semantic Import Versioning Compatibility
+
+Consider we have a module `foo` which defines the following `MsgDoSomething` and that we've released its state
+machine in go module `example.com/foo`:
+
+```protobuf
+package foo.v1;
+
+message MsgDoSomething {
+  string sender = 1;
+  uint64 amount = 2;
+}
+
+service Msg {
+  DoSomething(MsgDoSomething) returns (MsgDoSomethingResponse);
+}
+```
+
+Now consider that we make a revision to this module and add a new `condition` field to `MsgDoSomething` and also
+add a new validation rule on `amount` requiring it to be non-zero, and that following go semantic versioning we
+release the next state machine version of `foo` as `example.com/foo/v2`.
+
+```protobuf
+// Revision 1
+package foo.v1;
+
+message MsgDoSomething {
+  string sender = 1;
+  
+  // amount must be a non-zero integer.
+  uint64 amount = 2;
+  
+  // condition is an optional condition on doing the thing.
+  //
+  // Since: Revision 1
+  Condition condition = 3;
+}
+```
+
+Approaching this naively, we would generate the protobuf types for the initial
+version of `foo` in `example.com/foo/types` and we would generate the protobuf
+types for the second version in `example.com/foo/v2/types`.
```

Comment:
A specific major-versioned protobuf package e.g. foo.v1 should always corresponds to a single Go package e.g. example.com/foo/v1. If maintainers make a change to that specific major-versioned protobuf — which, as is correctly described here, must be backwards compatible — then that change must also be reflected in the corresponding application code. If changes in a protobuf definition were only satisfied by application code with a new identifier, that would defeat the purpose of using protobufs in the first place: they exist to manage this kind of change over time.

Reply:
Much of the discussion in the rest of the ADR seems to be based on this misunderstanding. I suspect a re-evaluation will produce different insights and outcomes.

Reply:
Thanks for your comments @peterbourgon. I think maybe a critical distinction that is necessary to understand is that state machines and protobuf APIs have effectively distinct major versions. State machines often have what we call state breaking changes but may support the same protobuf package at different revisions.

So when you say that a proper change to a protobuf package should never result in a different go package or different app code that is fundamentally not true for the SDK. It's not a misunderstanding on our part.

Reply:
(To be explicit, I'm assuming that we're talking about protobufs used to express data interchange between distinct processes over a network connection, as other use cases, e.g. as interchange format between entities in a single process, is nonsensical.)

> A critical distinction that is necessary to understand is that state machines and protobuf APIs have effectively distinct major versions. State machines often have what we call state breaking changes but may support the same major protobuf package at different revisions.

Behavior is part of the API contract established by protobufs, or any other IDL. That is, if you have some app code statemachine/v1 which implements protobuf foo.v1 in some way, and then you add statemachine/v2 which implements protobuf foo.v1 in a different and non-backwards-compatible way, this is invalid. You _must_ define a new protobuf foo.v2 with the new semantics.

Look at it from the client perspective. The client knows only what is part of the protocol negotiation, which abstracts over implementation. It doesn't know, by design, if the foo.v1 service it connects to is implemented by statemachine/v1 or statemachine/v2.

Reply:
> (To be explicit, I'm assuming that we're talking about protobufs used to express data interchange between distinct processes over a network connection, as other use cases, e.g. as interchange format between entities in a single process, is nonsensical.)

We are not at all concerned with network processes here, but rather communication between parts of code within the same process which you call nonsensical. However, it's not that nonsensical as we're not intending to marshal things, just use the same generated type definitions.

> That is, if you have some app code statemachine/v1 which implements protobuf foo.v1 in some way, and then you add statemachine/v2 which implements protobuf foo.v1 in a different and non-backwards-compatible way, this is invalid. You must define a new protobuf foo.v2 with the new semantics, even if the methods and types of the API are the same.

> Look at it from the client perspective. The client knows only what is part of the protocol negotiation, which abstracts over implementation. It doesn't know, by design, if the foo.v1 service it connects to is implemented by statemachine/v1 or statemachine/v2. All foo.v1 implementations must be equivalent in effect to each other.

We get all that believe me. But one, we're not talking about backwards incompatible changes. Everything we're taking about assumes the same basic semantics for a given API. But assuming the same basic semantics doesn't mean byte for byte state machine equivalence. Behavior can vary between state machines in ways that is compatible from the client perspective, but breaking from a blockchain determinism perspective. Also new features can be added and we need to be correctly forwards compatible which is actually not trivial.

Reply:
> We are not at all concerned with network processes here, but rather communication between parts of code within the same process . . .

Protocol buffers are not an appropriate solution to this class of problem.

Reply:
Why not?

Reply:
Actually I'd argue that having the module version compatibility we're trying to achieve here is almost intractable to do correctly without an IDL like protobuf

Reply:
> But assuming the same basic semantics doesn't mean byte for byte state machine equivalence. Behavior can vary between state machines in ways that is compatible from the client perspective, but breaking from a blockchain determinism perspective.

Consider a foo.v1 interface with a single method do_something(account). Consider two nodes N1 and N2 which each offer an API endpoint that implements that interface. Let N1's implementation withdraw 1000atom from the account, and let N2's implementation deposit 1000atom into the account.

A client sends a request to a load balancer that randomly chooses either N1 or N2 to serve the request. This is common and valid and acceptable, as both nodes advertise exactly the same API at the same version. Either choice succeeds, as both N1 and N2 provide implementations which are equivalent at the protobuf level, and compatible from the client perspective. But the random choice of the load balancer produces wildly different results. That dog ain't gon' hunt. If you implement foo.v1 you can't behave in any meaningfully different way to other implementations of foo.v1.

Reply:
> Why not?

I'm honestly a bit confused by this question. [Protocol buffers are](https://developers.google.com/protocol-buffers)

> Google's language-neutral, platform-neutral, extensible mechanism for **serializing structured data**

Serialization is necessary when data crosses a boundary where its representation is potentially ambiguous, like when you write a uint64 to a network connection (little endian or big endian?) or a protobuf struct to a file (how do you encode a string? how does the FS lay out individual blocks? etc.) 

But in a given memory space, this ambiguity doesn't exist. When you call a function in Go that takes an int64 or returns a struct { Name string, Value int } or whatever, the in-memory representation of those things is well-defined. You can just pass them on over, either as a value or a pointer, doesn't matter. Serialization serves no purpose here.

Reply:
> > But assuming the same basic semantics doesn't mean byte for byte state machine equivalence. Behavior can vary between state machines in ways that is compatible from the client perspective, but breaking from a blockchain determinism perspective.
> 
> Consider a foo.v1 interface with a single method do_something(account). Consider two nodes N1 and N2 which each offer an API endpoint that implements that interface. Let N1's implementation withdraw 1000atom from the account, and let N2's implementation deposit 1000atom into the account.
> 
> A client sends a request to a load balancer that randomly chooses either N1 or N2 to serve the request. This is common and valid and acceptable, as both nodes advertise exactly the same API at the same version. Either choice succeeds, as both N1 and N2 provide implementations which are equivalent at the protobuf level, and compatible from the client perspective. But the random choice of the load balancer produces wildly different results. That dog ain't gon' hunt. If you implement foo.v1 you can't behave in any meaningfully different way to other implementations of foo.v1.

We're not talking about that wide of a variation. A concrete use case would be distribution being compatible with both liquid staking and regular staking. In those cases, the API and behavior is similar enough but some of the mechanics are very different. But still for clients it's similar enough. Also, assuming you know this but in a blockchain scenario you could never have two nodes implementing different behavior. You could only have 2 networks with different behavior, ex. osmosis with liquid staking and cosmos hub with regular staking.

Reply:
> Serialization is not necessary when data moves within a single memory space.

This ADR isn't about serializing things and actually isn't really even about protobuf. Protobuf generated types happen to be the API boundary and are actually pretty convenient for that. But we're trying to avoid in memory serialzation at all costs of course.

Reply:
> Protobuf generated types happen to be the API boundary and are actually pretty convenient for that. But we're trying to avoid in memory serialzation at all costs of course.

This is so confusing. Protobufs are in no way meant to be used to express API boundaries within a single program or process; that problem is solved with language-native types. Protobufs are a tool for managing the serialization and deserialization of data between independent communicating entities over time. If you're not serializing, you shouldn't be using protobufs.

This is I hope not controversial stuff!

Reply:
In the context of this ADR, I'll observe that SDK modules are fundamentally Go packages, with Go APIs. It's fine to expose those APIs with protobuf services and protobuf generated types, but (a) the only reason to do that is to make them accessible over the network, and (b) protobufs are just one of many possible ways to do so. You could just as easily expose the API with some code that did JSON/RPC, or some kind of RESTy stuff, or over Avro, or with SOAP, or whatever. Protobufs are exactly equivalent to any of these things. None of them should "leak" their details into the module itself.

Reply:
Thanks for your efforts to understand this @peterbourgon. I think ADR 033 is helpful context for this ADR. Basically all modules use protobuf generated code to expose their services to the network. They can also use the same generated code and services to talk to other modules (usually without serialization). Currently modules are in the same process but we're imagining the boundary between them being sort of like a VM boundary. In the future, there may be an actual VM. It's kind of like hashicorp's go plug-in use of grpc which is also not over a network boundary.

Reply:
> Basically all modules use protobuf generated code to expose their services to the network. They can also use the same generated code and services to talk to other modules (usually without serialization)

A module may happen to expose its capabilities over the network with protobufs, that's fine, but that doesn't mean protobufs are an intrinsic part that module definition.

To ASCII it up, it's not

```
+--------------+
| protobuf     |
|  +--------+  |
|  | module |  |
|  +--------+  |
+--------------+
```

it's

```
+----------+ 
| protobuf | ... 
+----------+-----+
| module         |
+----------------+
```

and thus module-to-module communication in the same process shouldn't involve protobufs at all.

> (VM boundary)

Go's mandatory runtime + associated baggage (GC, etc.) makes it really really difficult to bridge thru the sort of FFI implied by terms like "VM" while maintaining reasonable efficiency. CosmWasm is a great example of that inefficiency: each call takes (last I checked) ca. ~10ms round-trip. Fine for a tech demo, nonviable for anything serious.

> hashicorp's go-plugin . . . which is also not over a network boundary

The hashicorp/go-plugin model definitely goes over a network boundary, it's just that the connection is typically a Unix socket on the same host. That's fine, the important distinction is that the RPC goes out of the process boundary, which means the data is serialized, which means protobufs make sense.

Reply:
Sorry if it's difficult to understand. I feel like it would make sense if you just looked at some of the module code we're talking about. As I've said there's no serialization involved  except in edge cases that we're trying to avoid. Types like Coin, MsgSend, etc. get generated from proto files. They are indeed part of the module to module boundary.

Reply:
> Types like Coin, MsgSend, etc. get generated from proto files. They are indeed part of the module API for both network and in process.

I understand that this is the case today for many types in the SDK. That's clear. I'm trying to explain that this is a serious problem that needs to be fixed, and not a pattern that can or should be extended any further.

Reply:
> I'm trying to explain that this is a problem that needs to be fixed, not a pattern that can or should be extended any further.

If that's your position then it would be good to see some rationale for it. First of all, it certainly is a pattern that _can_ be extended further, but if you think it _shouldn't_ then why? What specific things would you do differently and why would that be better?

Reply:
I've linked to the landing page for protobufs which explicitly describes it as "a mechanism for serializing structured data" and in fact further suggests that you think of it like a faster XML. Would you define a Coin as an XSLT schema? Would you use an XML code generator to produce the .go files that implement a Coin?

Is there any mainstream Go project which defines API boundaries _within a compilation unit_ via protobufs? That is, if some function in package A wants to call some function in package B, it can't just call B.function, it's expected to go through types that were defined by a protobuf file and code-generated by some process outside of the Go toolchain and versioned completely independently of the versioning scheme already established by the language itself?

Correct me if I'm wrong! If this is a reasonable pattern, where else is it used? Are there blog posts talking about the problems it solves? Conference talks that explain when and how it's appropriate to deploy? Is the pattern used in the standard library anywhere? Is it mentioned in any stdlib or golang.org/x or any similarly canonical package docs?

Reply:
Appreciate your engagement with this issue, but none of these seem like actual arguments. They all basically amount to "because other people aren't doing it then it shouldn't be done". But that's not a technical argument because technical arguments are about engineering tradeoffs. What are the actual engineering tradeoffs?


The proto files define the `Coin` type and therefore there will be a generated `Coin` struct. Why shouldn't we use that generated type for other things besides network communication? I don't care whether other people are doing it or not. I'm sure I could find some projects that are but that's not the point. I care whether or not it has benefits or drawbacks and whether those are the right tradeoffs to make.

Reply:
> Why shouldn't we use that generated type for other things besides network communication

I'm running out of ways to rephrase "because that is not what protocol buffers are for" so I guess I'll stop.

Reply:
Fair enough

Reply:
I think it makes sense to add an Approach D on "avoiding protobufs in public go APIs". I think there are ways the separate proto versioning makes things harder. Although I also think it's unavoidable to some extent because people want to send proto messages from modules to modules, but I think it's fair to propose limiting it which I gather is the gist of your argument.

Reply:
> people want to send proto messages from modules to modules

What do you mean by "modules" here? It seems like you might mean "Go modules"? That's reinforced by your statement

> "avoiding protobufs in public go APIs"

but this doesn't make sense. How could a public Go API have any concept of protobufs? Go doesn't understand protobufs, right? Not at the language, module, package, or API level. 

A Go API is authoritatively defined by the exported identifiers of the relevant Go code as it exists in the repo. It's fine if you want to use `protoc` to codegen some .go files, but unless you commit and push those generated .go files to your repo, they're not part of your module or package or public API. This makes the .proto definitions incidental. The only thing that matters is the Go code.

In any case, 

> people want to send proto messages from module to module

This doesn't make sense.

First, because modules define identifiers and not state, so nobody sends messages from a module (or package) to another module (or package), they create instances of values whose types are defined by modules (or packages) and link those values together. 

Second, because people don't want to send _proto_ messages between things, they want to send _information_ between things, and information is always natively expressed as Go types. Protobufs are one of many possible serializations of data between components, ultimately irrelevant to users. I don't send a cosmos.v1beta1.whatever.Coin.v1, I send a cosmos-sdk/types.Coin.

But ultimately it's the same misunderstanding we've been hammering on for I don't know how long. Protocol buffers are a serialization protocol, like XML or JSON or ASN.1. They are not a way to define types.

Reply:
> What do you mean by "modules" here? It seems like you might mean "Go modules"? That's reinforced by your statement

I mean SDK modules, not go modules.

> "avoiding protobufs in public go APIs"

What I mean is avoiding protobuf generated code in public go APIs so that the public API of go code is not subject to the separate versioning of protobuf.

> First, because modules define identifiers and not state, so nobody sends messages from a module (or package) to another module (or package), they create instances of values whose types are defined by modules (or packages) and link those values together.

No, I mean they want to send messages which were initially defined in protobuf from one SDK module to another SDK module.

> they want to send information between things, and information is always natively expressed as Go types

wrong, the protobuf API of the SDK has primacy over the golang API. It is the client-facing protocol and guaranteed to stay more constant than any golang APIs

> cosmos.v1beta1.whatever.Coin.v1, I send a cosmos-sdk/types.Coin

wrong, they want to send `cosmos.v1beta1.whatever.Coin.v1`

> But ultimately it's the same misunderstanding we've been hammering on for I don't know how long. Protocol buffers are a serialization protocol, like XML or JSON or ASN.1. They are not a way to define types.

That makes no sense. Protobuf does indeed define types. It's an IDL

Reply:
> wrong, the protobuf API of the SDK has primacy over the golang API. It is the client-facing protocol and guaranteed to stay more constant than any golang APIs
>
> Protobuf does indeed define types. It's an IDL

The protobuf API has primacy _on the wire_ but not _at the language level_. 

If I `go mod init dummy` and then `vim main.go` and `import "cosmos-sdk/types"` and then `var c types.Coin` then that `c` is valid. My program has no understanding of protobufs and doesn't need any understanding of protobufs. I can pass that `c` to any other function that takes a `cosmos-sdk/types.Coin` without involving protobufs at all. Protobufs do not enter the picture at all. 

Running `protoc` is the responsibility of the programmer who wants to define a protobuf adapter layer on top of some already-defined types. That programmer is obliged commit and maintain that generated code.

If I write a SDK module that has some Keeper method that takes a Coin, I need to import cosmos-sdk/types to get the Coin _and that's it_. My module factually does not need to know anything about any protobuf in order to use a Coin. 

Maybe if my module existed as a sidecar process that needed to expose a gRPC handler to be communicated-with, I'd need to serialize and deserialize Coins according to some proto definition of a Coin, and in that case I'd need to know about the protobuf definition of the Coin type. But that is _orthogonal and independent_ to my module's core business logic. If I don't do that, I don't need protobufs.

Reply:
> function that takes a cosmos-sdk/types.Coin without involving any notion of protobufs. Protobufs do not enter the picture at all.

`cosmos-sdk/types.Coin` is actually a [protobuf-generated type](https://github.com/cosmos/cosmos-sdk/blob/1ac7d3236e2301f476b5686e1640944381558026/types/coin.pb.go#L32) now in the SDK, so it does involve protobuf.

But I think I see your point. We _could_ do in the SDK something like manually write (i.e. not generate) business logic in Golang, like coins:

```go
// In cosmos-sdk/types/coin.go
type Coin struct {/* usual fields */}
```

Then generate protobuf code inside e.g. an internal package, and create To and From functions from the business logic Coin to the proto Coin. This is what Tendermint does, AFAIK, see examples of [To](https://github.com/tendermint/tendermint/blob/75d51e18f740c7cbfb7d8b4d49182ee6c7f41982/types/block.go#L220) and [From](https://github.com/tendermint/tendermint/blob/75d51e18f740c7cbfb7d8b4d49182ee6c7f41982/types/block.go#L242). This looks more like your ASCII art scenario 1.

In the SDK, we're abusing the proto-generated types to also serve as our business types and Golang public API. I think there are pros and cons. The big con I'm hearing from this thread is that protobuf is not meant for this. One pro of this approach is less boilerplate code (= potential source of bug). Another is that the SDK was criticized for routinely making (golang) API breaking changes, which was a headache for chains; switching to protobuf as the module boundary would give stronger API guarantees.

Anyways, this approach should definitely be explored, and from my understanding this will be in a new Approach D?

Reply:
In terms of what one module expects of another module, the protobuf API has primacy. Those type definitions are more guaranteed to be stable and define a public client API. What is exposed as keepers is seen as more unstable and there are even arguments for removing that layer in favor of just the protobuf API. I hear your counter arguments @peterbourgon. I think that's worth considering and will be added as approach D for considering pros and cons.

Reply:
@amaurym 

> switching to protobuf as the module boundary would give stronger API guarantees.

There is no fundamental difference in the guarantees provided by an API expressed as a set of identifiers exported by a Go package vs. an API expressed as a set of types and methods expressed as protobuf types and service definition(s). They're different, and they express invariants at separate conceptual levels. But protobufs are not "stronger" than Go package identifiers. Protobufs define an encoding format, nothing more.

—

@robert-zaremba 

> The current Cosmos SDK design is that proto types define the blockchain interface. If you make a transaction, you use that proto definitions.

I understand that modules currently express their exported Go language API identifiers via code that's generated from proto definitions. But this doesn't mean that the proto definition expresses the authoritative module API in any context except over the wire. Specifically, if I'm some Go code A compiled in the same process as some module B, I can talk to that module B without any knowledge of protobufs whatsoever. So "If I make a transaction" then _factually_ I don't use the proto definition, I use the generated code.

Reply:
> There is no fundamental difference in the guarantees provided by an API expressed as a set of identifiers exported by a Go package vs. an API expressed as a set of types and methods expressed as protobuf types and service definition(s).

Yes. But acknowledging Protobuf definition as a source of truth makes it more universal. In other words, we guarantee that other programs using objects defined using code coming from the same protobuf files are accepted (inter-process (so yes, wire format)).  

> But this doesn't mean that the proto definition expresses the authoritative module API in any context except over the wire. 

except when we say that the proposed Go API package is equivalent to the proto.

We can define API: 
1. using documentation, 
2. using language specific interfaces, 
3. or using more universal DSL, like protobuf, and unify that with the generated code.

I agree that using a wire format (which would be defined by an ABI) as an API is an anti pattern. But the concept here is to express API using Protobuf (again, I know that protobuf defines also the wire format) AND realize it by the generated code (so language level API).

Reply:
> Protobuf definition as a source of truth makes it more universal. In other words, we guarantee that other programs using objects defined using code coming from the same protobuf files are accepted (inter-process (so yes, wire format)).

Protobufs define types and services and other stuff at a conceptual layer _above_ the programming language which implements them. The mapping from proto definition to language code makes no compatibility guarantees. Two different .proto files with incompatible definitions can absolutely generate code which is incompatible. For example it's totally possible to write a foo.proto at rev 123 that generates a Go type X with definition X1 to whatever/foo.go, and for foo.proto at rev 124 to generate the same Go type X with definition X2 to the same whatever/foo.go, and for X1 and X2 to be entirely incompatible.

If rev 123 is tagged as e.g. v1.0.0 and rev 124 is tagged as e.g. v2.0.0, no problem! In this case the Go module version of each proto version's generated code is consistent with expected compatibility semantics. But nothing enforces this requirement. Nothing prevents rev 123 to be tagged as e.g. v1.0.0 and rev 124 to be tagged as v1.0.1. 

The concrete test case here is some consumer that imports your Go code at whatever version, and calls it directly from the same process. This is always possible, always valid, and by definition must abide the compatibility semantics defined by Go module/package versioning. Whatever may have been used to generate that Go code is irrelevant.

So if I have some code that imports your thing at v1.0.0 and calls some thing.Service with a ServiceRequest that happened to be codegen'd by a thing.proto at some point in time, right? And I build that code and resolves the Go module require of yourcompany.com/thing v1.0.0 to the actual v1.0.0 and everything works, yeah? And then later you update your thing.proto and change, say, the ServiceRequest in a way that's not backwards compatible with the previous version? And then codegen some .go code from that new .proto, and commit and push and tag that Go code as not v2.0.0 but as like v1.0.1? Well Go thinks that it's free to choose any version of yourcompany.com/thing as long as the major version is v1, and will treat all of those v1.*.* versions as essentially equivalent at compile time. Furthermore, unless you ensure (and enforce) changes to same-version proto definitions satisfy backwards compatibility, then diffs like this are by default gonna result in servers compiled for versions ...V crashing if they get requests from versions V+1 or above.

—

> But the concept here is to express API using Protobuf (again, I know that protobuf defines also the wire format) AND realize it by the generated code (so language level API).

This would be possible if protobuf definitions and language types had some kind of compatibility relationship to each other. But this is not generally true, it's only true when you include a mapping function between the two domains. And that mapping function has to be expressed explicitly in the translation, it can't be assumed or implicit. If this isn't obvious please let me know and I'll try to explain in more detail, but 🙅 until that's reflected in the ADR.

Reply:
Do you understand how we maintain proto version compatibility @peterbourgon both with buf breaking and the explicit guidelines documented in other ADRs? Usage of buf breaking is called out as an explicit assumption in this ADR and that invalidates your premise of minor revision incompatibility.

Reply:
> The mapping from proto definition to language code makes no compatibility guarantees. 

It can, that's why we are discussing the API package, which will reflect exactly the proto version.

> Nothing prevents rev 123 to be tagged as e.g. v1.0.0 and rev 124 to be tagged as v1.0.1.

It does. See Aaron comment above. 
BTW, you can use same argument for anything. If you make API breaking change in your library, technically you don't need to bump a major version, unless you have a team or tools which will prevent that.

> This would be possible if protobuf definitions and language types had some kind of compatibility relationship to each other. But this is not generally true, it's only true when you include a mapping function between the two domains.

That's why I was proposing correlating the proto version with the module version (so both the structural and logical versioning): https://github.com/cosmos/cosmos-sdk/pull/11802#discussion_r1029632886 and https://github.com/cosmos/cosmos-sdk/pull/11802#issuecomment-1330153995

Reply:
> That's why I was proposing correlate the proto version with the module version (so both the structural and logical versioning): https://github.com/cosmos/cosmos-sdk/pull/11802#discussion_r1029632886 and https://github.com/cosmos/cosmos-sdk/pull/11802#issuecomment-1330153995

I'm pretty sure this is equivalent to approach C) but with even less potential interoperability between module versions. With this approach, we'd either need to 100% use keepers with no protobuf types in the interfaces (approach D) or we would simply have no multi-version compatibility - i.e. every module is compatible with exactly one version of its dependent modules with no exceptions. So that doesn't accomplish any of the goals of this ADR, and that may be fine if we choose to abandon these goals, but I'm not sure which goals it does solve.

Also we rejected this as almost the worst case scenario for clients. I explained that somewhat here: https://github.com/cosmos/cosmos-sdk/discussions/10406#discussioncomment-1515050.

Reply:
> I'm pretty sure this is equivalent to approach C) 

Sort of. AFAIU, approach C is to keep "as is". Today modules don't have versions (except consensus version, which is yet another thing). Also, if `foo` depends on `bar/v1`, and there is `bar/v2` update which is not breaking what the `foo/v1` is using, then `foo/v1` can use both `bar/v1` and `bar/v2`.

Reply:
So sounds like you're saying `bar/v1` would correspond to the proto files in `bar.v1` and that `bar/v2` would correspond to the `bar.v2` proto files. Are you assuming that `bar/v2` will implement accept the `bar.v1` + `bar.v2` proto definitions and that `bar.v2` _only_ contains new definitions not in `bar.v1`, with `bar.v1` unchanged?